File: src/main/kotlin/at/fhhgb/mc/component/sim/positioning/StrongestAccessPointPositioning.java
package at.fhhgb.mc.component.sim.positioning;

import at.fhhgb.mc.component.sim.model.Position;
import at.fhhgb.mc.component.sim.model.recorder.network.AccessPoint;

import java.util.Collections;
import java.util.List;

/*
* Name: Michael Nigl
*
* ID: 1510455019
*/
public class StrongestAccessPointPositioning extends BasePositioning {


    public StrongestAccessPointPositioning(List<AccessPoint> accessPoints) {
        super(accessPoints);
    }


    @Override
    public Position calculatePosition(List<AccessPoint> accessPointList) {
        Collections.sort(accessPointList);
        Collections.reverse(accessPointList);
        for (AccessPoint accessPoint : accessPointList) {
            for (AccessPoint ap : getAccessPointList()) {
                if (ap.getBSSID().toLowerCase().equals(accessPoint.getBSSID().toLowerCase())) {
                    return new Position(ap.getPosition().getX(), ap.getPosition().getY());
                }
            }
        }
        return null;
    }
}

File: src/main/kotlin/at/fhhgb/mc/component/sim/positioning/StrongestAccessPointPositioning.kt
package at.fhhgb.mc.component.sim.positioning

import at.fhhgb.mc.component.sim.model.Position
import at.fhhgb.mc.component.sim.model.recorder.network.AccessPoint

/**
 * A Positioning method that uses the strongest access point (highest signal strength) for the position estimation.
 *
 * @property allAccessPoints a list of all installed access points in the building.
 *
 * @constructor Initializes a new Positioning Method
 */
class StrongestAccessPointPositioning(val allAccessPoints: List<AccessPoint>) : IPositioning {

    override fun calculatePosition(accessPointList: List<AccessPoint>): Position {
        for (scanned in accessPointList.sorted().reversed()) {
            allAccessPoints
                    .filter { it.bssid.toLowerCase() == scanned.bssid.toLowerCase() }
                    .firstOrNull()?.let {
                        return it.position
                    }
        }

        return Position(0, 0)
    }
}
